ovirt_imageio/client/_options.py

```diff
@@ -78,19 +140,54 @@ def config_file(self):
         return os.path.join(base_dir, "ovirt-img.conf")
 
     def _merge_config(self, args):
+        """
+        Read options from specified config section and set the value for unused
+        command line options. This must be done after parsing command line
+        options since we depend on --config.
+        """
         config = configparser.ConfigParser(interpolation=None)
         config.read([self.config_file])
 
         if not config.has_section(args.config):
             self._parser.error(
                 f"No section: '{args.config}' in '{self.config_file}'")
 
-        for name in ("engine_url", "username", "password", "cafile"):
-            if getattr(args, name) is not None:
+        for option in self._OPTIONS:
+            # Skip command line only options.
+            if not option.config:
+                continue
+
+            # Skip options set by the command line.
+            if getattr(args, option.name, None) is not None:
                 continue
 
-            if config.has_option(args.config, name):
-                setattr(args, name, config.get(args.config, name))
+            # Read and validate value from config file.
+            if config.has_option(args.config, option.name):
+                raw_value = config.get(args.config, option.name)
+                try:
+                    value = option.type(raw_value)
+                except ValueError as e:
+                    self._parser.error(e)
+                setattr(args, option.name, value)
+
+    def _set_defaults(self, args):
+        """
+        Add default value for unused command line arguemnt (None by default),
+        and unset config only values (attribute missing). This must be done
+        after reading config values.
+        """
+        for option in self._OPTIONS:
+            if getattr(args, option.name, None) is None:
+                setattr(args, option.name, option.default)
+
+    def _check_required_options(self, args):
+        """
+        Check that all required options have some value. This must be done
+        after setting the defaults.
+        """
+        for option in self._OPTIONS:
+            if option.required and getattr(args, option.name) is None:
+                self._parser.error(f"Option '{option.name}' is required")
```

Comment:
It's suboptimal that we have 2 levels of required-arguments-checking, instead of telling the user the whole list:

```
$ ./ovirt-img  download-disk 931f3b6-669a-46d2-8309-f23d0575cdaf 
usage: ovirt-img download-disk [-h] [-c CONFIG] [--engine-url ENGINE_URL]
                               [--username USERNAME]
                               [--password-file PASSWORD_FILE]
                               [--cafile CAFILE] [--log-file LOG_FILE]
                               [--log-level LOG_LEVEL] [-f {raw,qcow2}]
                               [--max-workers MAX_WORKERS]
                               [--buffer-size BUFFER_SIZE]
                               disk_id filename
ovirt-img download-disk: error: the following arguments are required: filename
$ ./ovirt-img  download-disk 931f3b6-669a-46d2-8309-f23d0575cdaf diskname
usage: ovirt-img [-h] {download-disk} ...
ovirt-img: error: Option 'engine_url' is required
```

But I think it's ok, he can do it step by step too.

Reply:
Right, it should be easy to fail with the all the missing arguments instead.

Reply:
Fixing the 2 levels of required arguments checking may be possible if we teach argparse to make disk_id and filename optional, and implement the check in our parser. But I don't have time for this now, we can improve this later based on feedback from users.

We show now all the missing required arguments instead of the first one.